galsim/second_kick.py

```diff
@@ -0,0 +1,174 @@
+# conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions, and the disclaimer given in the accompanying LICENSE
+#    file.
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions, and the disclaimer given in the documentation
+#    and/or other materials provided with the distribution.
+#
+"""@file second_kick.py
+
+This file implements the atmospheric PSF "second kick", which is an analytic calculation for the
+surface brightness profile of an atmospheric PSF convolved with an Airy PSF in the infinite exposure
+limit.  The atmospheric turbulence spectrum follows the von Karman formulation with an optional
+truncation scale.  Generally, this will be used in conjunction with a PhaseScreenPSF being drawn
+using geometric photon-shooting.  In this case, the PhaseScreenPSF will simulate the effects of the
+low frequency turbulence modes, which can be treated purely using refraction, while the SecondKick
+handles the high frequency modes."""
+
+import numpy as np
+
+import galsim
+
+from . import _galsim
+from .gsobject import GSObject
+
+
+class SecondKick(GSObject):
+    """Class describing the infinite exposure limit (or equivalently, the expected value) for the
+    surface brightness profile of an atmospheric PSF convolved by an Airy PSF.  The atmospheric PSF
+    is assumed to arise from a turbulent phase spectrum following a von Karman power spectrum of
+    fluctuations (See the VonKarman docstring for more details).
+
+    The intended use for this profile is as a correction to applying the geometric approximation to
+    PhaseScreenPSF objects when drawing using geometric photon shooting.  The geometric
+    approximation is only valid for length scales larger than some critical scale where the effects
+    of interference are unimportant.  For smaller length scales, interference (diffraction) must be
+    handled using an optical paradigm that acknowledges the wave nature of light, such as Fourier
+    optics.
+
+    Fourier optics calculations are many orders of magnitude slower than geometric optics
+    calculations, however, so we implement a scale-splitting algorithm first described in Peterson
+    et al. (2015) for the LSST PhoSim package.  Essentially, phase fluctuations below a critical
+    mode in Fourier space, labeled `kcrit`, are handled by the fast geometric optics calculations
+    present in PhaseScreenPSF.  Fluctuations for Fourier modes above `kcrit` are then calculated
+    analytically by SecondKick.  Because very many oscillations of these high-k modes both fit
+    within a given telescope aperture and pass by the aperture during a moderate length exposure
+    time, we can use the same analytic expectation value calculation for the high-k component of all
+    PSFs across a field of view, thus incurring the somewhat expensive calculation for Fourier
+    optics only once.
```

Comment:
I think at this point it would be helpful to describe the two limiting cases of SecondKick for kcrit=0 and kcrit=inf.  These limiting cases (which you test in test_second_kick.py) were very helpful for me in understanding what this class is really modeling.  In particular, that it always includes the full effect of the Airy profile, and just additionally includes some portion of the VonKarman profile, which depends on the kcrit value.